influxdb3_write/src/last_cache.rs

```diff
@@ -0,0 +1,2225 @@
+        cache_name: Option<&str>,
+        predicates: &[Predicate],
+    ) -> Option<Result<Vec<RecordBatch>, ArrowError>> {
+        self.cache_map
+            .read()
+            .get(db_name)
+            .and_then(|db| db.get(tbl_name))
+            .and_then(|tbl| {
+                if let Some(name) = cache_name {
+                    tbl.get(name)
+                } else if tbl.len() == 1 {
+                    tbl.iter().next().map(|(_, lc)| lc)
+                } else {
+                    None
+                }
+            })
+            .map(|lc| lc.to_record_batches(predicates))
+    }
+}
+
+/// A Last-N-Values Cache
+///
+/// A hierarchical cache whose structure is determined by a set of `key_columns`, each of which
+/// represents a level in the hierarchy. The lowest level of the hierarchy holds the last N values
+/// for the field columns in the cache.
+pub(crate) struct LastCache {
+    /// The number of values to hold in the cache
+    ///
+    /// Once the cache reaches this size, old values will be evicted when new values are pushed in.
+    count: LastCacheSize,
+    /// The time-to-live (TTL) for values in the cache
+    ///
+    /// Once values have lived in the cache beyond this [`Duration`], they can be evicted using
+    /// the [`remove_expired`][LastCache::remove_expired] method.
+    ttl: Duration,
+    /// The key columns for this cache
+    key_columns: Vec<String>,
+    /// The Influx Schema for the table that this cache is associated with
+    schema: Schema,
+    /// The internal state of the cache
+    state: LastCacheState,
+}
+
+impl LastCache {
+    /// Create a new [`LastCache`]
+    fn new(count: LastCacheSize, ttl: Duration, key_columns: Vec<String>, schema: Schema) -> Self {
+        Self {
+            count,
+            ttl,
+            key_columns,
+            schema,
+            state: LastCacheState::Init,
+        }
+    }
+
+    /// Push a [`Row`] from the write buffer into the cache
+    ///
+    /// If a key column is not present in the row, the row will be ignored.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if the internal cache state's keys are out-of-order with respect to the
+    /// order of the `key_columns` on this [`LastCache`]
+    pub(crate) fn push(&mut self, row: &Row) {
+        let mut target = &mut self.state;
+        let mut key_iter = self.key_columns.iter().peekable();
+        while let (Some(key), peek) = (key_iter.next(), key_iter.peek()) {
+            if target.is_init() {
+                *target = LastCacheState::Key(LastCacheKey {
+                    column_name: key.to_string(),
+                    value_map: Default::default(),
+                });
+            }
+            let Some(value) = row
+                .fields
+                .iter()
+                .find(|f| f.name == *key)
+                .map(|f| KeyValue::from(&f.value))
+            else {
+                // ignore the row if it does not contain all key columns
```

Comment:
We might want to represent null key column values in the structure? I'm not sure about this yet though. Something to think about and discuss.

Reply:
When I first refactored this, I had the `LastCacheKey` hold a `HashMap<Option<KeyValue>, LastCacheState>` to handle null key values, but switched to this behaviour to simplify it.

I think It just needs to do that, i.e., use `Option<KeyValue>` instead of `KeyValue`, and then store the datatype in the `LastCacheKey`, because 1) key columns will always have a fixed data type, and 2) you can't rely on the `KeyValue` alone to get the data type when it is `None` (for creating `RecordBatch`es).

I can open up an issue for this.

Reply:
Yeah, still not totally sure we want to bother with this, so maybe log an issue to see if anyone cares. For now I'd leave it as is.